### ISSUE TITLE
_NET_WM_FULLSCREEN_MONITORS not handled properly.

### ISSUE THREAD

**Issue by fabrox** [i3/i3#2116]
i3 is not handling the _NET_WM_FULLSCREEN_MONITORS (http://standards.freedesktop.org/wm-spec/wm-spec-latest.html#idm140200472531472) hint properly. There are several applications that use this hint to span a fullscreen window over several monitors.

One example is freerdp - If started with /multimon /f the resulting session window should span over all connected monitors. In i3 the resulting window is only shown in fullscreen on the monitor from which freerdp is started and thus the space on monitors to the top/left/right is invisible.

I don't know if that helps but some window managers that get it right are for example xfwm4, dwm and awesome.

i3 versions i tested it with are 4.11 and 4.10.4.

**Comment by i3bot** [comment]
I don’t see a link to logs.i3wm.org. Did you follow http://i3wm.org/docs/debugging.html? (In case you actually provided a link to a logfile, please ignore me.)

**Comment by i3bot** [comment]
I don’t see a version number. Could you please copy & paste the output of `i3 --version` into this issue?

**Comment by fabrox** [comment]
i3 version 4.11 (2015-09-30, branch "4.11")

**Comment by Airblader** [comment]
> i3 is not handling the _NET_WM_FULLSCREEN_MONITORS ([…]) hint properly.

i3 is not handling it all. However, i3 also doesn't claim to handle it and hence any application should be able to deal with that.

> There are several applications that use this hint to span a fullscreen window over several monitors.

i3 only allows to span a fullscreen container over one or all monitors, nothing in between anyway.

> I don't know if that helps but some window managers that get it right are for example xfwm4, dwm and awesome.

I've only looked at the awesome source code, but awesome doesn't handle this hint at all either.

**Comment by fabrox** [comment]
>> i3 is not handling it all. However, i3 also doesn't claim to handle it and hence any application should be >> able to deal with that.

Yes they should, absolutely. But unfortunately they don't and thus I've thought it would be easyer to get i3 
handling this hint than implementing "workarounds" for applications that make use of this hint.

>> i3 only allows to span a fullscreen container over one or all monitors, nothing in between anyway.

What do I need to tell i3 to span a window over all screens? I did not knew that it is possible and it this would solve my problem at least to a certain degree.

>> I've only looked at the awesome source code, but awesome doesn't handle this hint at all either. That >>suggests that it might be something other than this atom that is relevant for your use case.
>>
>> How did you come to the conclusion that it's about support of this atom?

I came to this conclusion because it worked fine with freerdp until they decided to get rid of the override_redirect approach and instead use NET_WM_STATE_FULLSCREEN and NET_WM_FULLSCREEN_MONITORS. 

They actually fixed it broken with this commit for several other window managers too and are working on  a "legacy mode" currently so override_redirect would be used in case NET_WM_FULLSCREEN_MONITORS is not supported by the wm.

But, however, I think handling it properly or handling it at all would be a nice feature to i3.

**Comment by Airblader** [comment]
Can you link to the commit that broke it there? That'd be quite interesting. 

You can fullscreen a container in i3 over all monitors by issuing 

    fullscreen toggle global

**Comment by Airblader** [comment]
> Yes they should, absolutely. But unfortunately they don't 

Just to make this clearer: the specification has a protocol of how a window manager tells applications which atoms they support. Unsupported non-mandatory atoms MUST be dealt with by the client application. It's not just a "they should"; they are violating the spec and it should be a clear bug against them. 

That isn't to say we can't support this atom. I just want to be clear on the fact that this is an obvious bug on the client side.

**Comment by fabrox** [comment]
I think this is the merged pull request in question: https://github.com/FreeRDP/FreeRDP/pull/2426 and this is the approach to fix it https://github.com/FreeRDP/FreeRDP/pull/2821

**Comment by Airblader** [comment]
Thanks. The PR is indeed broken since, as stated in the fix PR, it doesn't check for window manager support.

As far as supporting the atom in i3, if I understand it correctly, it would allow clients to fullscreen them on arbitrary sets of outputs. That would be a problem for us since we don't currently support that (and I doubt that it's trivial). And only half-supporting an atom might cause more headache than it does good.

Let's see what @stapelberg says, though.

**Comment by stapelberg** [comment]
In principle, I’m not opposed to implementing this hint, since I can understand the use-case. However:

From http://standards.freedesktop.org/wm-spec/wm-spec-latest.html#idm140200472531472:

> A read-only list of 4 monitor indices indicating the top, bottom, left, and right edges of the window when the fullscreen state is enabled. The indices are from the set returned by the Xinerama extension.

I find it a bit problematic that Xinerama is required here. i3 in general strives to use RandR for everything, and adding a dependency on Xinerama doesn’t sound appealing.

Further, the standard doesn’t say anything about whether the window manager is allowed to refuse fullscreening on certain monitors, and what the behavior should be when the window cannot be positioned on the provided monitor indices (e.g. because they have a gap or their coordinates do not align).

So, there are three things to clarify before we can make a decision:

1. Is there a way to obtain the monitor indices from RandR instead? If not, it might make sense to add a hint to the EWMH spec which uses RandR indices.
2. Can we clarify my questions with regards to what the standard currently doesn’t describe? What do apps typically do/assume? How do other window managers implement this hint?
3. Can we update the EWMH spec to be more clear, in case we decided to go forward with an implementation after clarifying points 1 and 2?

**Comment by Airblader** [comment]
Neither awesome nor openbox deal with this atom at all. XFCE's window manager supports it, but only by means of setting it when a client message is received, in which case the Xinerama indices are given in the message (by the client).

This all really makes me think that it's not worth supporting this atom and it also seems like an application breaking due to this should reconsider its implementation since it'll be broken in several major window managers (at the least).

**Comment by stapelberg** [comment]
SGTM — let’s push for a fix in the application.

**Comment by StevenPolley** [comment]
I really hope commenting on this doesn't re-open the issue, sorry if it does.  I just wanted to mention that anyone with this problem, I have a crappy workaround but it at least works...

When you open xfreerdp with /multimon and /f, hit ctrl + alt + delete, ensure the rdp window is selected, then press right CTRL to escape input from RDP so you can issue i3 commands, then press mod + shift + space.  The window is now floating.  You can move the window with mod and left mouse to position it correctly.  

That's all assuming you have fairly default key bindings.  Kind of sucks having to do it each time.  I'm sure this could be automated using an i3 config for this type of window class.

**Comment by Airblader** [comment]
> are there any workarounds for this?

There shouldn't be an issue globally fullscreening an application from within i3 if by screens you mean outputs. See comments above.

**Comment by AlexKornitzer** [comment]
@Airblader can't seem to get multiple monitors to work in VirtualBox either at all. You can assign them to a different display but nothing happens. It seems like it tries to send it to the other display but can't. Unless I am being super dumb here.

**Comment by AlexKornitzer** [comment]
@Airblader Looks like @stapelberg's comment is correct - can we not answer those question and get this atom supported. This lacking atom must be the cause for the above as mentioned in the spec:
```
...
 Virtual machine software may use this hint to have a virtual operating system instance that sees multiple monitors. The application window stretches over several monitors, giving the appearance that these monitors have been taken over by the guest virtual machine. 
...
```

**Comment by Airblader** [comment]
> can we not answer those questions

Sure, anyone who's interested can go ahead and answer them. :-) However, I wouldn't hold my breath for changes to this in the EWMH, especially considering that even if you get a RandR-based protocol for this you'd still have to implement it in VirtualBox. Due to Wayland I think this has a 0% chance of happening, but if it does, we'd likely be talking of at least a couple years before you have this working on all ends.

By the way, the original post of this issue states that it works in awesome, which I already verified to not be supporting this atom, so I think the assessment that freerdp relies on this atom and nothing else is probably incorrect. It might make sense to look into what protocol is actually used under window managers supporting this as it doesn't seem to be `_NET_WM_FULLSCREEN_MONITORS`, at least not for freerdp.

I did also just check VirtualBox's code and this is the relevant part:

```
/* This method tests whether the current X11 window manager supports full-screen mode as we need it.
 * Unfortunately the EWMH specification was not fully clear about whether we can expect to find
 * all of these atoms on the _NET_SUPPORTED root window property, so we have to test with all
 * interesting window managers. If this fails for a user when you think it should succeed
 * they should try executing:
 * xprop -root | egrep -w '_NET_WM_FULLSCREEN_MONITORS|_NET_WM_STATE|_NET_WM_STATE_FULLSCREEN'
 * in an X11 terminal window.
 * All three strings should be found under a property called "_NET_SUPPORTED(ATOM)". */
/* static */
bool VBoxGlobal::supportsFullScreenMonitorsProtocolX11()
{
    /* Using a global to get at the display does not feel right, but that is
     * how it is done elsewhere in the code. */
    Display *pDisplay = QX11Info::display();
    Atom atomSupported            = XInternAtom(pDisplay, "_NET_SUPPORTED",
                                                True /* only_if_exists */);
    Atom atomWMFullScreenMonitors = XInternAtom(pDisplay,
                                                "_NET_WM_FULLSCREEN_MONITORS",
                                                True /* only_if_exists */);
    Atom atomWMState              = XInternAtom(pDisplay,
                                                "_NET_WM_STATE",
                                                True /* only_if_exists */);
    Atom atomWMStateFullScreen    = XInternAtom(pDisplay,
                                                "_NET_WM_STATE_FULLSCREEN",
                                                True /* only_if_exists */);
    bool fFoundFullScreenMonitors = false;
    bool fFoundState              = false;
    bool fFoundStateFullScreen    = false;
    Atom atomType;
    int cFormat;
    unsigned long cItems;
    unsigned long cbLeft;
    Atom *pAtomHints;
    int rc;
    unsigned i;

    if (   atomSupported == None || atomWMFullScreenMonitors == None
        || atomWMState == None || atomWMStateFullScreen == None)
        return false;
    /* Get atom value: */
    rc = XGetWindowProperty(pDisplay, DefaultRootWindow(pDisplay),
                            atomSupported, 0, 0x7fffffff /*LONG_MAX*/,
                            False /* delete */, XA_ATOM, &atomType,
                            &cFormat, &cItems, &cbLeft,
                            (unsigned char **)&pAtomHints);
    if (rc != Success)
        return false;
    if (pAtomHints == NULL)
        return false;
    if (atomType == XA_ATOM && cFormat == 32 && cbLeft == 0)
        for (i = 0; i < cItems; ++i)
        {
            if (pAtomHints[i] == atomWMFullScreenMonitors)
                fFoundFullScreenMonitors = true;
            if (pAtomHints[i] == atomWMState)
                fFoundState = true;
            if (pAtomHints[i] == atomWMStateFullScreen)
                fFoundStateFullScreen = true;
        }
    XFree(pAtomHints);
    return fFoundFullScreenMonitors && fFoundState && fFoundStateFullScreen;
}

/* static */
bool VBoxGlobal::setFullScreenMonitorX11(QWidget *pWidget, ulong uScreenId)
{
    return XXSendClientMessage(QX11Info::display(),
                               pWidget->window()->winId(),
                               "_NET_WM_FULLSCREEN_MONITORS",
                               uScreenId, uScreenId, uScreenId, uScreenId,
                               1 /* Source indication (1 = normal application) */);
}
```

What strikes me as interesting is the actual call to fullscreen as it sends the same `uScreenId` for all edges. I don't see how that would possibly cause the window to span *all* monitors with regards to the specification; seems like it'd only fullscreen on a single monitor. But perhaps I'm missing something?

**Comment by Airblader** [comment]
My concern with supporting this atom stands, as spanning arbitrary outputs is something i3 doesn't support and which is quite hard to support.

I wonder what would happen if you just added the atom to the `_NET_SUPPORTED` list manually, though. VirtualBox would think it's supported, the actual call would of course be ignored by i3. However, if you then use `fullscreen global`, I feel like it should still work (?)

**Comment by AlexKornitzer** [comment]
@Airblader thanks for looking into this. My interest is for supporting VMware Workstation rather than VirtualBox. I downloaded VB to see if this was just an issue with workstation but VB seems to display the same symptoms as freerdp it seems, only taking the screen it is loaded on.

If the decision to use Xinerama was made, does the issue then become handling the new layout in i3? I assume it is not as simple as telling i3 to handle a number of monitors as a single workspace?

Sorry for the stupid questions, I am not familiar with these sorts of code bases or protocols at all.

XFCE4 seems to be supporting it and before using xinerama used gtk: https://github.com/xfce-mirror/xfwm4/search?q=NET_WM_FULLSCREEN_MONITORS&type=Commits&utf8=%E2%9C%93

I can verify that the gtk method (with workstation) worked as I used to use xfce4 on my host before i3.

**Comment by Airblader** [comment]
> My interest is for supporting VMware Workstation rather than VirtualBox.

Since VMWare isn't open source I cannot check what they use and we generally do not attempt too hard to support closed-source software, so VirtualBox would be the strongest argument here to make changes.

> If the decision to use Xinerama was made […]

i3 already supports Xinerama. But Xinerama is outdated and limited (e.g., it does not support adding or removing outputs at runtime) and thus we want to avoid it. In fact, we currently only use Xinerama if RandR is absent (which should virtually never be the case these days) or if you force it via config, and then we print a big warning telling you to avoid doing it at all cost.

One specific problem with supporting this atom would be that it only works if you run i3 in Xinerama mode (which as stated above we really recommend you don't), meaning we'd have to add the atom to `_NET_SUPPORTED` conditionally based on what mode i3 is run with. That all sounds really bad to me. I don't see a good reason why VirtualBox (or VMWare) cannot achieve the same thing using RandR.

**Comment by AlexKornitzer** [comment]
Could we support the atom a similar way to xfce4 before they updated to support xinerama indices as shown in my previous link? I am happy to have a stab at getting this working with VirtualBox and then tackling Workstation, just have no idea where to start :)

**Comment by Airblader** [comment]
> Could we support the atom a similar way to xfce4 before they updated to support Xinerama indices as shown in my previous link?

I'm not sure I follow; that change in xfce4 was a bugfix because their implementation was simply broken. The answer to such a suggestion is certainly no, we will not implement it knowingly broken to avoid looking up Xinerama information. :-)

> I am happy to have a stab at getting this working with VirtualBox and then tackling Workstation, just have no idea where to start :)

For starters, why not try what I mentioned above? Get i3 to claim the atom is supported (which is I think a one-line change) and then try to use `fullscreen global`.

**Comment by AlexKornitzer** [comment]
Of course, was just thinking if they got away with using gtk and the only issue was primary layout then there should be a way to replicate the correct indices without Xinerama.

Anyway, let me try your suggestion first and then I can go from there :)

**Comment by AlexKornitzer** [comment]
Right added in the atom (with no additional changes) and Workstation is now playing ball, it successfully takes over both monitors.

I suppose the next step is getting the workspace too extend across the two monitors? I am thinking something like this:

```
3 workspaces and when 2 is fullscreen and resides on primary monitor:
 _______ _______
|       |       |
|   1   |   3   |
|_______|_______|
 primary  secondary

 _______ _______
|       |       |
|   2   | 2 ext |
|_______|_______|
 primary  secondary
```

Apologies for the terrible diagram, hopefully that makes sense!

**Comment by Airblader** [comment]
Getting i3 to span workspaces across monitors could be done either using Xinerama or the (test-only and not supported) internal `fake_outputs` directive. Admittedly, this becomes quite hack-ish at this point.

**Comment by AlexKornitzer** [comment]
So is the functionality obtained just be adding the atom the best possible at the moment?

Rather than spanning the workspace I was thinking more like emulating it, such that 2ext is a workspace but is not accessible. e.g. 2ext is effectively a dead workspace. So switching to workspace 3 would also swap screen 1 to workspace 1 rather than leaving it on 2.

**Comment by Airblader** [comment]
> So is the functionality obtained just be adding the atom the best possible at the moment?

I think so, yes.

**Comment by AlexKornitzer** [comment]
Can we get this atom into master then?

When I find some time, i'll have a play with implementing my idea above.

**Comment by Airblader** [comment]
> Can we get this atom into master then?

No, because we're not handling it, so that'd be a lie. :-)

**Comment by AlexKornitzer** [comment]
I would argue just having it as basic support as is, is pretty useful, been using it all day today with no issues. But fair enough, guess I'll have to implement handling it first then :P

**Comment by Airblader** [comment]
> I would argue just having it as basic support as is, is pretty useful

Maybe so, but it isn't spec-compliant. We don't know whether it'll break some other application which, rightfully, expects us to actually implement the atom if we claim to support it in `_NET_SUPPORTED`.

> guess I'll have to implement handling it first then :P

Which as discussed above is entirely unclear as to whether we even want to support this in i3.

**Comment by AlexKornitzer** [comment]
So can we make a decision on this? If I managed to get a POC that is working without Xinerama would it still be a no?

**Comment by AlexKornitzer** [comment]
From what I understand all that Xinerama is providing, is the x,y,width,height of the monitor, surely this can be done without Xinerama even if the spec demands it. Or is the issue that `XineramaQueryScreens` could return a different order to a different solution? Even if this is the case surely a helper can be used to replicate the order through arithmetic on the x,y,w,h?

**Comment by Airblader** [comment]
But the atom uses Xinerama indices to tell the window manager how to fullscreen the window. So we need to understand these indices which per specification requires Xinerama. I don't see how arithmetic could solve this. What exactly a Xinerama index represents depends on how it is configured.

We certainly don't want to rely on "could"s. The spec requires Xinerama, we don't want to use Xinerama, hence we cannot use this atom. As pointed out much earlier in this thread, we'd prefer the specs to be updated here. Alternatively, clients should use different mechanisms.

**Comment by alexeicolin** [comment]
The workaround for VMware by @AlexKornitzer also works for Awesome WM: add `_NET_WM_FULLSCREEN_MONITORS` to `atoms.list` and to `atom` array in `ewmh_init` in `ewmh.c`.

PS. I apologize for spaming i3, and hope you can donate this comment space for those landing here looking for a workaround.

**Comment by marksolaris** [comment]
Here is the patch required to add to fvwm2 2.6.7 so that VMWare Workstation 14 can use the "Cycle multiple monitors" button and your VM can span screens. Note the backend isn't implemented by this patch, it just advertizes the _NET_WM_FULLSCREEN_MONITORS atom which is enough to trick Workstation.

`
diff -Nbaur fvwm-2.6.7.orig/fvwm/ewmh.c fvwm-2.6.7/fvwm/ewmh.c
--- fvwm-2.6.7.orig/fvwm/ewmh.c 2016-10-18 22:54:47.000000000 +1100
+++ fvwm-2.6.7/fvwm/ewmh.c  2018-02-28 13:35:13.575868365 +1100
@@ -146,6 +146,9 @@
    ENTRY(
        "_NET_WM_STATE_FULLSCREEN",      XA_ATOM,
        ewmh_WMStateFullScreen),
+   ENTRY(
+       "_NET_WM_FULLSCREEN_MONITORS",      XA_ATOM,
+       ewmh_WMStateFullScreen),
    ENTRY("_NET_WM_STATE_HIDDEN",          XA_ATOM,   ewmh_WMStateHidden),
    ENTRY(
        "_NET_WM_STATE_MAXIMIZED_HORIZ", XA_ATOM,
diff -Nbaur fvwm-2.6.7.orig/tests/hints/hints_test.c fvwm-2.6.7/tests/hints/hints_test.c
--- fvwm-2.6.7.orig/tests/hints/hints_test.c    2016-09-30 23:54:33.000000000 +1000
+++ fvwm-2.6.7/tests/hints/hints_test.c 2018-02-28 13:44:35.784866221 +1100
@@ -41,6 +41,8 @@
 Atom ATOM_NET_WM_STATE_HIDDEN          = None;
 Atom ATOM_NET_WM_STATE_STAYS_ON_TOP    = None;
 Atom ATOM_NET_WM_STATE_FULLSCREEN      = None;
+/* Fiction */
+Atom ATOM_NET_WM_FULLSCREEN_MONITORS      = None;
 
 Atom ATOM_NET_WM_DESKTOP = None;
 
@@ -173,6 +175,8 @@
    ATOM_NET_WM_STATE_HIDDEN          = XIA("_NET_WM_STATE_HIDDEN");
    ATOM_NET_WM_STATE_FULLSCREEN      = XIA("_NET_WM_STATE_FULLSCREEN");
    ATOM_NET_WM_STATE_STAYS_ON_TOP    = XIA("_NET_WM_STATE_STAYS_ON_TOP");
+   /* Fiction */
+   ATOM_NET_WM_FULLSCREEN_MONITORS      = XIA("_NET_WM_FULLSCREEN_MONITORS");
 
    ATOM_NET_WM_DESKTOP               = XIA("_NET_WM_DESKTOP");
 
@@ -192,7 +196,7 @@
    printf("      all, border, resizeh, title, menu, minimize, maximize\n");
    printf("  --ewmh-state <names>\n");
    printf("      hidden, shaded, sticky, skippager, skiptaskbar,\n");
-   printf("      maxhoriz, maxvert, modal, staysontop, fullscreen\n");
+   printf("      maxhoriz, maxvert, modal, staysontop, fullscreen, monitors\n");
    printf("  --ewmh-type <names>\n");
    printf("      normal, dock, toolbar, desktop, menu, dialog, splash, utility\n");
    printf("  --mwm-input { modless, app_modal, sys_modal, full_app_modal}\n");
@@ -553,19 +557,13 @@
            {
                states[state_count++] =
                    ATOM_NET_WM_STATE_MODAL;
-           }
-           else if (strcasecmp(argv[i], "staysontop") == 0)
-           {
-               states[state_count++] =
-                   ATOM_NET_WM_STATE_STAYS_ON_TOP;
-           }
-           else if (strcasecmp(argv[i], "fullscreen") == 0)
-           {
-               states[state_count++] =
-                   ATOM_NET_WM_STATE_FULLSCREEN;
-           }
-           else
-           {
+           } else if (strcasecmp(argv[i], "staysontop") == 0) {
+               states[state_count++] = ATOM_NET_WM_STATE_STAYS_ON_TOP;
+           } else if (strcasecmp(argv[i], "fullscreen") == 0) {
+               states[state_count++] = ATOM_NET_WM_STATE_FULLSCREEN;
+           } else if (strcasecmp(argv[i], "monitors") == 0) {
+               states[state_count++] = ATOM_NET_WM_FULLSCREEN_MONITORS;
+           } else { 
                error_arg = "--ewmh-state";
            }
        }
`

**Comment by CyberShadow** [comment]
mpv's documentation claims that it relies on window managers implementing this hint to provide the `--fs-screen` option:

```
       --fs-screen=<all|current|0-32>
              In  multi-monitor  configurations  (i.e.  a  single  desktop that spans
              across multiple displays), this option tells mpv  which  screen  to  go
              fullscreen  to.   If default is provided mpv will fallback on using the
              behavior depending on what the user provided with the screen option.

                 Note (X11)

                        This option does works properly  only  with  window  managers
                        which understand the EWMH _NET_WM_FULLSCREEN_MONITORS hint.
```

I think it would be good to support this hint even without the ability to span multiple arbitrary monitors.

**Comment by psychon** [comment]
@CyberShadow https://github.com/i3/i3/issues/2116#issuecomment-165705565 (basically: this cannot be implemented without loosing other (and IMHO more important) features)

**Comment by jbbr** [comment]
This missing feature also causes multi monitor support in "Citrix Receiver" to not work:

```bash
❯ /opt/Citrix/ICAClient/wfica -span h somedesktop-connection.ica
1: 1920x1200 at 0, 02: 1920x1080 at 1920, 0Window manager does not support multi-monitor windows.
```
I think this information might be interesting for people looking for i3<>citrix issues.

**Comment by Chandru-B** [comment]
> This missing feature also causes multi monitor support in "Citrix Receiver" to not work:
> 
> ```shell
> ❯ /opt/Citrix/ICAClient/wfica -span h somedesktop-connection.ica
> 1: 1920x1200 at 0, 02: 1920x1080 at 1920, 0Window manager does not support multi-monitor windows.
> ```
> 
> I think this information might be interesting for people looking for i3<>citrix issues.
Did you manage to get this working? My Virtualbox spans 2 i3 windows even with a single monitor. Not sure how to get this working

**Comment by jbbr** [comment]
@Chandru-B no it's still broken for me.
If I want to use Citrix Receiver/Workspace with multimonitor support I switch to XFCE/Gnome/... Not sure who to blame, this is basically the only thing I dislike when using i3 - Hope either Citrix or i3 will change their implementation to get this working sometime.
Would be nice to have some kind of workaround for this issue.

**Comment by Chandru-B** [comment]
@jbbr Thanks. 
I have been living by toggling, Got used to the following now.
1)Ctrl+F2 once ica launches
2) toggle to full screen on one monitor using below
 bindsym $mod+f fullscreen toggle
3) switch off my laptop screen(another shortcut in i3 config)
4) Toggle to 2 desktop monitors with the following shortcut
 bindsym $mod+Shift+f        fullscreen toggle global

**Comment by zhou13** [comment]
Is it possible to add the workaround from @AlexKornitzer to master? That will at least make my life easier. I still need to use VMware to do the presentation.  If this does not break any known software, ...

**Comment by Airblader** [comment]
You can just patch your i3 locally. I do not intend to merge a workaround that makes i3 break spec to fix a broken client.

**Comment by zhou13** [comment]
Sure: https://aur.archlinux.org/packages/i3-gaps-fullscreen-next-git/

I guess we need to install i3 with more and more suffixes.

**Comment by Airblader** [comment]
@zhou13

> todo

Mentioned in this issue was freerdp, which has since been fixed (https://github.com/FreeRDP/FreeRDP/pull/4319), so perhaps you should just update. If this affects a different application for you, I would suggest focusing your effort on getting that fixed rather than appending "more and more suffixes" to i3. Of course if you don't at least report to the application in question you won't ever see it fixed.

**Comment by zhou13** [comment]
The software I had a problem is VMware. Since it is not opensource, I don't think fixing on that end is easier than appending "more and more suffixes" to i3 :)

**Comment by Airblader** [comment]
Even VMware has a way to report bugs[1], it's just not much of an actual issue tracker. It doesn't change the fact that if you don't even alert them of their problem, it will never be fixed and you're robbing yourself of the chance to remove workaround patches again (which may well also become hard to maintain at some point).

Of course reporting the bug won't immediately fix it, maybe they won't fix it in the next year even, but it costs you a couple of minutes and at least gives you — and everyone else — a chance. You can still proceed to apply the patch for it afterwards.

[1] https://www.vmware.com/support/policies/defect.html

**Comment by zhou13** [comment]
https://communities.vmware.com/message/2864358#2864358

**Comment by tbnorth** [comment]
As of Citrix client 2311 I couldn't get it to work *at all* in i3, so here's my work around that got it working again

https://gist.github.com/tbnorth/510e4db4ab68fd82eb070331f22f9737
